notes/overriding.md

```diff
@@ -0,0 +1,675 @@
+expressive yet simple set of features upon which others can be built as
+libraries. However, in the case of Nix, overriding is a fundamental pattern, and
+having it implemented in user code leads to some well-known issues:
+
+- Definition, update or field access may not be as ergonomic as with vanilla
+  native records.
+- Several competing mechanisms have been implemented in slightly different
+  contexts. Overriding in general is already difficult for beginners, but having
+  to learn several ones, all similar but still different, is hard.
+- User land implementations don't have a first-class access to information such
+  as source location and metavalues. In a dynamic language, this can make good
+  error reporting much harder.
+- It is also potentially harder to make user land implementations efficient.
+
+See [this gist](https://gist.github.com/edolstra/29ce9d8ea399b703a7023073b0dbc00d)
+for more details. We continue with an overview of existing mechanisms in Nix and
+related languages.
+
+### Nixpkgs overriding
+
+The basic idea is to represent a recursive record explicitly as a function from
+a `self` record to the final record:
+
+```nix
+r = rec {
+  a = 1;
+  b = a + 1;
+}
+# r is represented as:
+rRepr = self: {
+  a = 1;
+  b = self.a + 1;
+}
+```
+
+`self` is a self-reference, in the style of `this` in object oriented languages.
+It is computed as a
+[fixpoint](https://en.wikipedia.org/wiki/Fixed-point_combinator), simply
+realized by auto-application thanks to laziness:
+
+```
+r = let fixpoint = rRepr fixpoint; in fixpoint
+```
+
+The explicit dependency to `self` gives the ability to provide a different
+definition for `self.a` in the final value. Overriding is achieved exactly by
+functionally updating the `a` field of the `self` parameter before passing it to
+the original representation:
+
+```nix
+let extension = {a = 2;}; in
+# The fixpoint of result is { a = 2; b = 3; }
+resultRepr = self: (rRepr (self // extension)) // extension
+```
+
+The second outer update ensures that the final result is also set to `a = 2`,
+and not only the `a` appearing in `b`.
+
+Some details are left out, but this is the gist of it. See also
+[this pill on overriding](https://nixos.org/guides/nix-pills/override-design-pattern.html)
+or [this article on fixpoints in Nix](http://r6.ca/blog/20140422T142911Z.html).
+
+**limits**
+
+- Use a specific representation, rather than handling good old plain records
+  (although the actual representation in Nixpkgs is more ergonomic than a plain
+  function).
+- Overriding nested attribute sets is painful. If one do the naive thing,
+  the whole subrecord is erased:
+  ```nix
+  let rRepr = self: {
+    a = {b = self.a.c;};
+  }; in
+  let extension = {a = {c = 2;};}; in
+  rExt = let fixpoint = rRepr (fixpoint // extension); in
+  fixpoint // extension
+  # Gives {a = {c = 2;};} instead of expected {a = {b = 2; c = 2;};}
+  ```
+- They don't [compose well](https://github.com/NixOS/nixpkgs/issues/16394).
```

Comment:
This is a bit of a fud-type sentence. It doesn't bear much meaning. Maybe you can say an extra sentence (like “Nixpkgs has been known to have several incompatible overriding mechanism”). Or maybe just “There are more known difficulties”.

Reply:
It is not about several mechanisms, but about applying several overrides to the same configuration. It seems the Nixpkgs basic overriding mechanism doesn't support this well, and may reset initial default values that were changed at some point as long as the last override doesn't update them (https://github.com/NixOS/nixpkgs/issues/16394#issuecomment-233032852). I should include a stand-alone example.

Reply:
After trying on simple examples, it doesn't seem to be a problem actually. Let me gather more info and come back to this.